### DISCUSSION TITLE
Static Indexers

### DISCUSSION THREAD

**Discussion by osamakawish** [dotnet/csharplang#4341] [Language Ideas]
This is just a convenience and I can easily work around this, but it cleans up the code a bit.

Basically it arises from situations where I collect all initiations of a given type into a list, then want to index it.

```c#
public class T {
    private static List<T> Ts { get; set; }
    public static T this[int i] => Ts[i];
}
```

It's an oversimplified example. And it can already be worked around via `public static T Get(int i)` or `public static void Set(int i, T value)` methods. But a minor suggestion that helps in some contexts.

### Usage

```c#
T[5].DoStuff();
```

I know it looks ridiculous given the way the indexer is defined via a this statement and all, but it's a suggestion. Maybe the notation can be changed though, to something like:

```c#
public static T T[int i] => Ts[i];
```

**Comment by huoyaoyuan** [comment]
I wonder if there are any potential syntax ambiguous.

There is also slight inconsistency. You declared the indexer with `this`, but you cannot call it with `this`.

Allowing parameterized property in C# (explicitly naming the indexer) gives more clear semantics.

[Static named indexer](https://sharplab.io/#v2:DYLgbgRgPgkgtgBwPYCcAuBnABAZQJ4ZoCmcAsAFAAKArhMAJYDGWAwsAIYbYsVZ9Y06TXAAt2KIgBMBKJAiLo8WGMTgYAFPQB2kogA8sAQWwwtxAOYKAlEZNmillL34uA4kTTOX3rACUP1ChaWNq6BgDUWACMXt4AojpY7p7kPnw4HrFpWFn8CdIZKS75MnIKaHgUJWycGEA===) is supported in VB and IL, but cannot be represented in C#.

**Comment by svick** [comment]
Previous discussion: https://github.com/dotnet/csharplang/issues/526.
